packages/client/tests/functional/methods/upsert/native-atomic/tests.ts

```diff
@@ -0,0 +1,241 @@
+      expect(doesSQLContainOnConflict).toEqual(true)
+    })
+
+    test('should only use ON CONFLICT when there is only 1 unique field in the where clause', async () => {
+      const name = faker.name.firstName()
+      const title = faker.name.jobTitle()
+
+      const user = await client.user.create({
+        data: {
+          name,
+          posts: {
+            create: {
+              title,
+            },
+          },
+        },
+      })
+
+      await expect(() =>
+        // This will fail
+        client.user.upsert({
+          where: {
+            // Because two unique fields are used
+            id: user.id,
+            name,
+          },
+          create: {
+            name,
+          },
+          update: {
+            name,
+          },
+        }),
+      ).rejects.toThrowError('Argument where of type UserWhereUniqueInput needs exactly one argument')
+
+      const sqlPromise: Promise<string> = createSQLPromise(client)
+      let doesSQLContainOnConflict = false
+
+      // This 'will' use ON CONFLICT
+      await client.user.upsert({
+        where: {
+          // Because only one unique field is used
+          name,
+        },
+        create: {
+          name,
+        },
+        update: {
+          name,
+        },
+      })
+      doesSQLContainOnConflict = (await sqlPromise).includes('ON CONFLICT')
+      expect(doesSQLContainOnConflict).toEqual(true)
+    })
+
+    test('should only use ON CONFLICT when the unique field defined in where clause has the same value as defined in the create arguments', async () => {
+      const name = faker.name.firstName()
+      const title = faker.name.jobTitle()
+
+      await client.user.create({
+        data: {
+          name,
+          posts: {
+            create: {
+              title,
+            },
+          },
+        },
+      })
+
+      let sqlPromise: Promise<string> = createSQLPromise(client)
+      let doesSQLContainOnConflict = false
+
+      // This will 'not' use ON CONFLICT
+      await client.user.upsert({
+        where: {
+          name,
+        },
+        create: {
+          // Because the the where 'name' is 'not' equal to the create 'name'
+          name: name + '1',
+        },
+        update: {
+          name: name + '1',
+        },
+      })
+      doesSQLContainOnConflict = (await sqlPromise).includes('ON CONFLICT')
+      expect(doesSQLContainOnConflict).toEqual(false)
+
+      sqlPromise = createSQLPromise(client)
+      // This 'will' use ON CONFLICT
+      await client.user.upsert({
+        where: {
+          name,
+        },
+        create: {
+          // Because the the where 'name' is 'equal' to the create 'name'
+          name,
+        },
+        update: {
+          name: name + '1',
+        },
+      })
+      doesSQLContainOnConflict = (await sqlPromise).includes('ON CONFLICT')
+      expect(doesSQLContainOnConflict).toEqual(true)
```

Comment:
Not directly related to the two lines I comment on here:

Do we have a simple way to add a snapshot of the full SQL query being executed for all these? I think it would be valuable to surface the full query here. (Of is our `faker` usage really dynamic and hence this would need to be sanitized?)

Reply:
I decided not to snapshot the query because it's different for each provider, and so you then need to switch the query or dump it into a separate file 'miles' away from the test itself. I think if we want to be snapshotting the entire SQL we should look more into https://github.com/prisma/prisma/pull/13981 .